src/librbd/internal.cc

```diff
@@ -3826,4 +3835,212 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
+
+    ImageCtx *imctx = new ImageCtx(image_name, "", "", image_ioctx, true);
+    r = imctx->state->open();
+    if (r < 0) {
+      lderr(cct) << "error opening image: "
+        	 << cpp_strerror(-r) << dendl;
+      delete imctx;
+      return r;
+    }
+
+    ldout(cct, 20) << "removing image " << image_name
+                   << " image id " << imctx->header_oid << dendl;
+
+    r = cls_client::group_image_add(&group_ioctx, group_header_oid,
+				    imctx->id, image_ioctx.get_id());
+
+    if (r < 0) {
+      lderr(cct) << "couldn't put image into removing state: "
+        	 << cpp_strerror(-r) << dendl;
+      goto exit;
+    }
+
+    r = cls_client::group_image_remove_uplink(&image_ioctx, imctx->header_oid,
+					      group_id, group_ioctx.get_id());
+    if ((r < 0) && (r != -ENOENT)) {
+      lderr(cct) << "couldn't remove group reference from image"
+        	 << cpp_strerror(-r) << dendl;
+      goto exit;
+    }
+
+    r = cls_client::group_image_remove(&group_ioctx, group_header_oid, imctx->id,
+				       image_ioctx.get_id());
+    if (r < 0) {
+      lderr(cct) << "couldn't remove image from group"
+        	 << cpp_strerror(-r) << dendl;
+      goto exit;
+    }
+    r = 0;
+
+exit:
+    imctx->state->close();
+    return r;
+  }
+
+  int group_image_list(librados::IoCtx& group_ioctx,
+		       const char *group_name,
+		       std::vector<std::tuple<std::string,
+					      int64_t,
+					      int>>& images)
+  {
+    CephContext *cct = (CephContext *)group_ioctx.cct();
+    ldout(cct, 20) << "group_list_images " << &group_ioctx
+                   << " group name " << group_name << dendl;
+
+    string group_id;
+
+    int r = cls_client::dir_get_id(&group_ioctx, RBD_GROUP_DIRECTORY, group_name, &group_id);
+    if (r < 0) {
+      lderr(cct) << "error reading consistency group id object: "
+                 << cpp_strerror(r)
+        	 << dendl;
+      return r;
+    }
+    string group_header_oid = util::group_header_name(group_id);
+
+    ldout(cct, 20) << "listing images in group name "
+                   << group_name << " group id " << group_header_oid << dendl;
+
+    std::vector<std::tuple<std::string, int64_t, int64_t>> image_ids;
+    r = cls_client::group_image_list(&group_ioctx, group_header_oid, image_ids);
+
+    if (r < 0) {
+      lderr(cct) << "error reading image list from consistency group: "
+        	 << cpp_strerror(-r) << dendl;
+      return r;
+    }
+
+    for (auto i : image_ids) {
+      librados::Rados rados(group_ioctx);
+      IoCtx ioctx;
+      rados.ioctx_create2(std::get<1>(i), ioctx);
+      std::string image_name;
+      r = cls_client::dir_get_name(&ioctx, RBD_DIRECTORY,
+                                   std::get<0>(i), &image_name);
+      if (r < 0) {
+        return r;
+      }
+      images.push_back(std::make_tuple(image_name,
+                                       std::get<1>(i),
+                                       std::get<2>(i)));
```

Comment:
Safely convert from private API state enum to public state (see how mirror state handles it)